src/dist/replication/meta_server/meta_bulk_load_service.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2017-present, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#include <dsn/dist/fmt_logging.h>
+
+#include "meta_bulk_load_service.h"
+
+namespace dsn {
+namespace replication {
+
+bulk_load_service::bulk_load_service(meta_service *meta_svc, const std::string &bulk_load_dir)
+    : _meta_svc(meta_svc), _bulk_load_root(bulk_load_dir)
+{
+    _state = _meta_svc->get_server_state();
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::initialize_bulk_load_service()
+{
+    task_tracker tracker;
+    error_code err = ERR_OK;
+
+    create_bulk_load_root_dir(err, tracker);
+    tracker.wait_outstanding_tasks();
+
+    if (err == ERR_OK) {
+        try_to_continue_bulk_load();
+    }
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::create_bulk_load_root_dir(error_code &err, task_tracker &tracker)
+{
+    blob value = blob();
+    _meta_svc->get_remote_storage()->create_node(
+        _bulk_load_root,
+        LPC_META_CALLBACK,
+        [this, &err, &tracker](error_code ec) {
+            if (ERR_OK == ec || ERR_NODE_ALREADY_EXIST == ec) {
+                ddebug_f("create bulk load root({}) succeed", _bulk_load_root);
+                sync_apps_bulk_load_from_remote_stroage(err, tracker);
+            } else if (ERR_TIMEOUT == ec) {
+                dwarn_f("create bulk load root({}) failed, retry later", _bulk_load_root);
+                tasking::enqueue(
+                    LPC_META_STATE_NORMAL,
+                    nullptr,
+                    std::bind(&bulk_load_service::create_bulk_load_root_dir, this, err, tracker),
+                    0,
+                    std::chrono::milliseconds(1000));
+            } else {
+                err = ec;
+                dfatal_f(
+                    "create bulk load root({}) failed, error={}", _bulk_load_root, ec.to_string());
+            }
+        },
+        value,
+        &tracker);
```

Comment:
```suggestion
    _meta_svc->get_meta_storage()->create_node(
        std::string(_bulk_load_root),
        std::move(value),
        [this, &err, &tracker](error_code ec) {
            ddebug_f("create bulk load root({}) succeed", _bulk_load_root);
            sync_apps_bulk_load_from_remote_stroage(err, tracker);
        });
```

Reply:
I don't use `get_meta_storage`, I should not use meta service's tracker here. I will try to create bulk load dir on remote storage, and if there is bulk load info existed, I will sync bulk load info synchronously like code below.
```
create_bulk_load_root_dir(err, tracker);
tracker.wait_outstanding_tasks();
```

Reply:
You can create a meta storage on your own if so, because `meta_storage` is simply a wrapper over `meta_state_service`: https://github.com/XiaoMi/rdsn/blob/master/src/dist/replication/meta_server/meta_state_service_utils.h#L49

```
task_tracker tracker;
meta_storage mstore(_meta_svc->get_remote_storage(), &tracker);
```

Reply:
I know it, and I just think it is unnecessary to write another wrapper class to implementation sync_bulk_load_from_zk function, in normal bulk load process, I still use `get_meta_storage`.